torch/csrc/autograd/python_legacy_variable.cpp

```diff
@@ -0,0 +1,126 @@
+#include "python_variable.h"
+
+#include <ATen/ATen.h>
+
+#include "torch/csrc/Exceptions.h"
+#include "torch/csrc/autograd/python_function.h"
+#include "torch/csrc/autograd/python_variable.h"
+#include "torch/csrc/tensor/python_tensor.h"
+
+using namespace at;
+
+namespace torch { namespace autograd {
+
+static PyObject *THPVariable_pynew(PyTypeObject* type, PyObject *args, PyObject *kwds) {
+  HANDLE_TH_ERRORS
+  THPObjectPtr _data;
+  PyObject *data = nullptr;
+  PyObject *grad_fn = nullptr;
+  char is_volatile = 0;
+  char requires_grad = 0;
+  const char* name = nullptr;
+
+  const char *accepted_args[] = {"data", "requires_grad", "volatile", "_grad_fn", "name", nullptr};
+  if (!PyArg_ParseTupleAndKeywords(args, kwds, "|ObbOz", (char**)accepted_args,
+      &data, &requires_grad, &is_volatile, &grad_fn, &name))
+    return nullptr;
+
+  if (grad_fn == Py_None)
+    grad_fn = nullptr;
+
+  if (is_volatile) {
+    PyErr_WarnEx(PyExc_UserWarning,
+        "volatile was removed and now has no effect. Use `with torch.no_grad():` "
+        "instead.", 1);
+  }
+
+  if (is_volatile && requires_grad) {
+    throw ValueError("Variable can't be volatile and require_grad at the same time!");
+  }
+  if (grad_fn && !THPFunction_Check(grad_fn)) {
+    throw TypeError("_grad_fn has to be a Function object or None, but got %s",
+        Py_TYPE(grad_fn)->tp_name);
+  }
+  Tensor tensor;
+  if (!data || data == Py_None) {
+    // For legacy serialization code, create an empty tensor. This is also used
+    // by nn.Parameter() with no arguments.
+    auto var = torch::tensor::get_default_tensor_type().tensor();
+    tensor = static_cast<Variable&>(var).data();
+  } else if (THPVariable_Check(data)) {
+    tensor = ((THPVariable*)data)->cdata.data();
+  } else {
+    throw torch::TypeError("Variable data has to be a tensor, but got %s",
+        Py_TYPE(data)->tp_name);
+  }
+
+  Variable var;
+  if (grad_fn) {
+    auto grad_fn_ = THPFunction_asFunction((THPFunction*)grad_fn);
+    Edge edge(grad_fn_, grad_fn_->bump_inputs());
+    var = make_variable(std::move(tensor), std::move(edge));
```

Comment:
I guess it's okay to pass requires_grad=False if there is a grad_fn? -- it just assumes you meant true?

Reply:
Yeah a non-empty edge implies `var.requires_grad() == true`. `make_variable` has two constructors now, so if you pass an edge you can't pass `requires_grad`:

```
make_variable(at::Tensor data, bool requires_grad = false);
make_variable(at::Tensor data, Edge gradient_edge);
```